### ISSUE TITLE
Overriding generic methods gives false positives/negatives?

### ISSUE THREAD

**Issue by patrick-kidger** [microsoft/pyright#5083]
**To Reproduce / expected behaviour**

```python
from abc import ABC, abstractmethod
from typing import Generic, TypeVar

T = TypeVar("T")

class A(ABC, Generic[T]):
  @abstractmethod
  def f(self, x: T) -> T:
    ...

# False negative: int/str skew means T is bound inconsistently
# Expected behaviour: an error.
class B(A[int]):
  def f(self, x) -> str:
    return "not an integer"

# False negative? int/str skew; the str is inferrable, although it is unannotated.
# I'm not sure about this one. Expected behaviour is an error if this is in-scope for pyright.
class C(A[int]):
  def f(self, x):
    return "not an integer"

# At least one of D1 and D2 are an error, depending on what is intended behaviour.

# This gives Operator "+" not supported for types "T@A" and "Literal[1]"
class D1(A[int]):
  def f(self, x):
    return x + 1

# This raises no error.
class D2(A[int]):
  def f(self, x: str):
    return x + "hi"

# If it is required to annotate overriden methods, then D2 is an error, due to the int/str skew
#     for TypeVar T.
# If it is not required, then D1 is an error, as T is bound to int (via `A[int]`), so `x: T` should
#     resolve to `x: int`.
```

**VS Code extension or command-line**
Pyright 1.1.302 from the command line.

**Additional context**
I think of abstract generics as being a very common thing to do, so I have a hunch that these must have been hit before, and that you're going to say that some or all of these are "as designed"!

Nonetheless I looked through the previously-closed issues / google / python typing docs / etc. and couldn't find any discussion of this.

In any case, all of the examples above seem like footguns, and I'm not sure what the correct thing to do here is.

**Comment by erictraut** [comment]
Pyright reports incompatible method overrides if you enable the `reportIncompatibleMethodOverride` rule. This is off by default for "basic" type checking but on by default in "strict" mode.

**Comment by patrick-kidger** [comment]
Marvellous, that works perfectly. Thank you for your help!
